docs/devel/kubectl-conventions.md

```diff
@@ -135,6 +135,77 @@ Updated: 8/27/2015
 * Use "TYPE" for the particular flavor of resource type accepted by kubectl, rather than "RESOURCE" or "KIND"
 * Use "NAME" for resource names
 
+## Command structure conventions
+
+For every command there should be a `NewCmd<CommandName>` function that creates the command and returns a pointer to a `cobra.Command`, which can later be added to other parent commands to compose the structure tree. There should also be a `<CommandName>Options` struct with a variable to every flag and argument declared by the command (and any other variable required for the command to run). This makes tests and mocking easier. The struct ideally exposes three functions:
+
+* `Complete`: Completes the struct variables with values that may not be directly provided, for example, by flags pointers. Here you will usually take the `args` slice and set the values as appropriate variables, instantiate configs or clients, etc.
+* `Validate`: performs validation and returns errors.
+* `Run<CommandName>`: runs the actual command, taking as assumption that the struct is complete with all required values to run, and they are valid.
+
+Sample command skeleton:
+
+```go
+// MineOptions contains all the options for running the mine cli command
+type MineOptions struct {
+  mineLatest bool
+}
+
+const (
+  mineLong = `Some long description
+for my command.`
+
+  mineExample = `  # Run my command's first action
+  $ kubectl mine first
+
+  # Run my command's second action on latest stuff
+  $ kubectl mine second --latest`
+)
+
+// NewCmdMine implements the kubectl mine command.
+func NewCmdMine(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+  opts := &MineOptions{}
+
+  cmd := &cobra.Command{
+    Use:     "mine [--latest]",
+    Short:   "Run my command",
+    Long:    mineLong,
+    Example: mineExample,
+    Run: func(cmd *cobra.Command, args []string) {
+      if err := opts.Complete(f, cmd, args, out); err != nil {
+        cmdutil.CheckErr(err)
+      }
+      if err := opts.Validate(); err != nil {
+        cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
+      }
+      if err := opts.RunMine(); err != nil {
+        cmdutil.CheckErr(err)
+      }
+    },
+  }
+  
+  cmd.Flags().BoolVar(&options.mineLatest, "latest", false, "Use latest stuff")
+  return cmd
+}
+
+// Complete completes all the required options for mine.
+func (o *MineOptions) Complete(f *cmdutil.Factory, cmd *cobra.Command, args []string, out io.Writer) error {
+  return nil
+}
+
+// Validate validates all the required options for mine.
+func (o MineOptions) Validate() error {
```

Comment:
Do any commands obey this convention? create, for example, does not. 
```
func ValidateArgs(cmd *cobra.Command, args []string) error
func RunCreate(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *CreateOptions) error
```

It would be useful to have one Builder and one generator example we could point to as models to follow.